sympy/stats/stochastic_process_types.py

```diff
@@ -274,7 +293,10 @@ def __call__(self, time):
         if not time.is_symbol and time not in self.index_set:
             raise IndexError("%s is not in the index set of %s"%(time, self.symbol))
         func_obj = Function(self.symbol)(time)
-        pspace_obj = StochasticPSpace(self.symbol, self, self.distribution)
```

Comment:
What if we would have just called `self.distribution()`.

Reply:
We still need to return `RandomIndexedSymbol`, so the `pspace_obj` needs to be created. Also, there is no implementation of `distribution()` method for `MarkovChain`, as I explained in my previous comment. So the error won't be resolved for `MarkovChain` Processes unless we pass their specific distribution methods as the 3rd argument to `RandomIndexedSymbol` (The 3rd argument would just remain `None` otherwise, which was causing the error)

So I think it is necessary to proceed this way, or did you have any other method in mind?

Reply:
What about returning an object of abstract `Distribution` class?
```diff
diff --git a/sympy/core/tests/test_args.py b/sympy/core/tests/test_args.py
index d33b261750..6f9c984123 100644
--- a/sympy/core/tests/test_args.py
+++ b/sympy/core/tests/test_args.py
@@ -1050,6 +1050,10 @@ def test_sympy__stats__crv__SingleContinuousPSpace():
     from sympy.stats.crv import SingleContinuousPSpace
     assert _test_args(SingleContinuousPSpace(x, nd))
 
+@SKIP("abstract class")
+def test_sympy__stats__rv__Distribution():
+    pass
+
 @SKIP("abstract class")
 def test_sympy__stats__crv__SingleContinuousDistribution():
     pass
diff --git a/sympy/stats/compound_rv.py b/sympy/stats/compound_rv.py
index da3fe28250..b2fad26abf 100644
--- a/sympy/stats/compound_rv.py
+++ b/sympy/stats/compound_rv.py
@@ -1,6 +1,6 @@
 from sympy import Basic, Sum, Dummy, Lambda, Integral
 from sympy.stats.rv import (NamedArgsMixin, random_symbols, _symbol_converter,
-                        PSpace, RandomSymbol, is_random)
+                        PSpace, RandomSymbol, is_random, Distribution)
 from sympy.stats.crv import ContinuousDistribution, SingleContinuousPSpace
 from sympy.stats.drv import DiscreteDistribution, SingleDiscretePSpace
 from sympy.stats.frv import SingleFiniteDistribution, SingleFinitePSpace
@@ -120,7 +120,7 @@ def conditional_space(self, condition, *, compound_evaluate=True, **kwargs):
         return new_pspace.conditional_space(condition)
 
 
-class CompoundDistribution(Basic, NamedArgsMixin):
+class CompoundDistribution(Distribution, NamedArgsMixin):
     """
     Class for Compound Distributions.
 
diff --git a/sympy/stats/crv.py b/sympy/stats/crv.py
index 07e8440dd8..9a4faff4ae 100644
--- a/sympy/stats/crv.py
+++ b/sympy/stats/crv.py
@@ -21,7 +21,7 @@
 from sympy.core.sympify import _sympify
 from sympy.external import import_module
 from sympy.stats.rv import (RandomDomain, SingleDomain, ConditionalDomain, is_random,
-        ProductDomain, PSpace, SinglePSpace, random_symbols, NamedArgsMixin)
+        ProductDomain, PSpace, SinglePSpace, random_symbols, NamedArgsMixin, Distribution)
 
 
 class ContinuousDomain(RandomDomain):
@@ -140,7 +140,7 @@ def set(self):
                 "Set of Conditional Domain not Implemented")
 
 
-class ContinuousDistribution(Basic):
+class ContinuousDistribution(Distribution):
     def __call__(self, *args):
         return self.pdf(*args)
 
diff --git a/sympy/stats/drv.py b/sympy/stats/drv.py
index a037386503..e5a1f09750 100644
--- a/sympy/stats/drv.py
+++ b/sympy/stats/drv.py
@@ -6,7 +6,7 @@
 from sympy.stats.crv import reduce_rational_inequalities_wrap
 from sympy.stats.rv import (NamedArgsMixin, SinglePSpace, SingleDomain,
                             random_symbols, PSpace, ConditionalDomain, RandomDomain,
-                            ProductDomain)
+                            ProductDomain, Distribution)
 from sympy.stats.symbolic_probability import Probability
 from sympy.sets.fancysets import Range, FiniteSet
 from sympy.sets.sets import Union
@@ -16,7 +16,7 @@
 from sympy.external import import_module
 
 
-class DiscreteDistribution(Basic):
+class DiscreteDistribution(Distribution):
     def __call__(self, *args):
         return self.pdf(*args)
 
diff --git a/sympy/stats/frv.py b/sympy/stats/frv.py
index ef24efaf45..98f2cc8499 100644
--- a/sympy/stats/frv.py
+++ b/sympy/stats/frv.py
@@ -19,7 +19,7 @@
 from sympy.sets.sets import FiniteSet
 from sympy.stats.rv import (RandomDomain, ProductDomain, ConditionalDomain,
                             PSpace, IndependentProductPSpace, SinglePSpace, random_symbols,
-                            sumsets, rv_subs, NamedArgsMixin, Density)
+                            sumsets, rv_subs, NamedArgsMixin, Density, Distribution)
 from sympy.external import import_module
 
 
@@ -181,7 +181,7 @@ def set(self):
     def as_boolean(self):
         return FiniteDomain.as_boolean(self)
 
-class SingleFiniteDistribution(Basic, NamedArgsMixin):
+class SingleFiniteDistribution(Distribution, NamedArgsMixin):
     def __new__(cls, *args):
         args = list(map(sympify, args))
         return Basic.__new__(cls, *args)
diff --git a/sympy/stats/joint_rv.py b/sympy/stats/joint_rv.py
index b2a6bc87d0..abbbcaa991 100644
--- a/sympy/stats/joint_rv.py
+++ b/sympy/stats/joint_rv.py
@@ -20,8 +20,9 @@
 from sympy.matrices import ImmutableMatrix, matrix2numpy, list2numpy
 from sympy.stats.crv import SingleContinuousDistribution, SingleContinuousPSpace
 from sympy.stats.drv import SingleDiscreteDistribution, SingleDiscretePSpace
-from sympy.stats.rv import (ProductPSpace, NamedArgsMixin,
-                            ProductDomain, RandomSymbol, random_symbols, SingleDomain, _symbol_converter)
+from sympy.stats.rv import (ProductPSpace, NamedArgsMixin, Distribution,
+                            ProductDomain, RandomSymbol, random_symbols,
+                            SingleDomain, _symbol_converter)
 from sympy.utilities.misc import filldedent
 from sympy.external import import_module
 
@@ -232,7 +233,7 @@ def _sample_pymc3(cls, dist, size):
     'numpy': SampleJointNumpy
 }
 
-class JointDistribution(Basic, NamedArgsMixin):
+class JointDistribution(Distribution, NamedArgsMixin):
     """
     Represented by the random variables part of the joint distribution.
     Contains methods for PDF, CDF, sampling, marginal densities, etc.
@@ -306,7 +307,7 @@ def __getitem__(self, key):
 
 
 
-class MarginalDistribution(Basic):
+class MarginalDistribution(Distribution):
     """
     Represents the marginal distribution of a joint probability space.
 
diff --git a/sympy/stats/matrix_distributions.py b/sympy/stats/matrix_distributions.py
index 4307b2dc64..8f945ed997 100644
--- a/sympy/stats/matrix_distributions.py
+++ b/sympy/stats/matrix_distributions.py
@@ -4,7 +4,7 @@
                             MatrixSymbol, MatrixBase, Transpose, MatrixSet,
                             matrix2numpy)
 from sympy.stats.rv import (_value_check, RandomMatrixSymbol, NamedArgsMixin, PSpace,
-                            _symbol_converter, MatrixDomain)
+                            _symbol_converter, MatrixDomain, Distribution)
 from sympy.external import import_module
 
 
@@ -153,7 +153,7 @@ def _sample_pymc3(cls, dist, size):
 #-------------------------Matrix Distribution----------------------------------#
 ################################################################################
 
-class MatrixDistribution(Basic, NamedArgsMixin):
+class MatrixDistribution(Distribution, NamedArgsMixin):
     """
     Abstract class for Matrix Distribution
     """
diff --git a/sympy/stats/rv.py b/sympy/stats/rv.py
index d869d216a6..f4a02f640f 100644
--- a/sympy/stats/rv.py
+++ b/sympy/stats/rv.py
@@ -349,6 +349,9 @@ def __new__(cls, symbol, n, m, pspace=None):
     symbol = property(lambda self: self.args[0])
     pspace = property(lambda self: self.args[3])
 
+class Distribution(Basic):
+    pass
+
 class ProductPSpace(PSpace):
     """
     Abstract class for representing probability spaces with multiple random
diff --git a/sympy/stats/stochastic_process_types.py b/sympy/stats/stochastic_process_types.py
index d1d05fe565..175f43740a 100644
--- a/sympy/stats/stochastic_process_types.py
+++ b/sympy/stats/stochastic_process_types.py
@@ -20,7 +20,7 @@
 from sympy.stats.joint_rv_types import JointDistributionHandmade
 from sympy.stats.rv import (RandomIndexedSymbol, random_symbols, RandomSymbol,
                             _symbol_converter, _value_check, pspace, given,
-                           dependent, is_random, sample_iter)
+                           dependent, is_random, sample_iter, Distribution)
 from sympy.stats.stochastic_process import StochasticPSpace
 from sympy.stats.symbolic_probability import Probability, Expectation
 from sympy.stats.frv_types import Bernoulli, BernoulliDistribution, FiniteRV
@@ -171,7 +171,7 @@ def state_space(self) -> tUnion[FiniteSet, Range]:
         return self.args[1]
 
     def distribution(self, key):
-        return None
+        return Distribution()
 
     def __call__(self, time):
         """
```
```
commit 72cf12b7af59bd0074075381a18538773370dfdd (HEAD -> issue_20078)
Author: Sidharth Mundhra <sidharthmundhra16@gmail.com>
Date:   Fri Dec 11 16:23:43 2020 +0530

    aims to fix simplification bug (#20078) with Wiener Process

commit ca83a68dd0b5e46a4e0d88ae9c1542f07affe782
Merge: e5c901e5f3 b1782d68f1
Author: Oscar Benjamin <oscar.j.benjamin@gmail.com>
Date:   Fri Dec 11 09:23:48 2020 +0000

    Merge pull request #19777 from theHamsta/__iter__
    
    Replace hasattr(... '__iter__') by check with `iterable`

commit e5c901e5f30bcc9bd0cb91f03d0696815b96bd1a
Merge: 87d967e81d 1fa5396d1e
Author: Oscar Benjamin <oscar.j.benjamin@gmail.com>
```
```python
In [1]: from sympy.stats import *

In [2]: T = Matrix([[0.5, 0.2, 0.3],[0.2, 0.5, 0.3],[0.2, 0.3, 0.5]])

In [3]: Y = DiscreteMarkovChain("Y", [0, 1, 2], T)

In [4]: simplify(Y[2]**2 + Y[2]**3)
Out[4]: 
               2
(Y[2] + 1)⋅Y[2] 
```